doc/developer/design/20230903_avro_doc.md

```diff
@@ -0,0 +1,390 @@
+deterministic—i.e., that future changes to comments on the columns and types
+used by the sink will *not* result in Materialize generating a different schema
+on restart.
+
+#### Catalog
+
+No changes to the catalog are required. The documentation will be automatically
+stored in the catalog via the addition of the `DOC ON` option, as the catalog
+stores the raw DDL statement for each sink.
+
+#### Avro schema generation
+
+The `interchange::json::build_row_schema_json` will need to be adjusted to
+take the requested documentation as a parameter. Loosely, it could grow a
+`docs: BTreeMap<String, String>` parameter, where the key represents the
+field specifier and the value represents the requested documentation for that
+field.
+
+The sink code in the coordinator would need to convert the `DOC ON` comments
+to this `BTreeMap`.
+
+## Alternatives
+
+### User-provided Avro schemas
+
+Materialize could instead allow users to *provide* an Avro schema with all
+required `doc` fields. For example, users could pre-publish the desired schema
+to the schema registry, and then Materialize could load that schema at `CREATE
+SINK` time.
+
+There are several issues with this approach:
+
+  * It is not ergonomic. The provided schema must exactly match the schema
+    Materialize generates, *except* for the `doc` fields. Minor errors in
+    constructing the schema (e.g., using a `long` where an `int` is required, or
+    ordering fields wrong) will result in hard to debug failures.
+
+  * It prevents Materialize from making backwards compatible changes to sink
+    schemas.
+
+    Today, we can make changes the generated Avro sink schemas as long as they
+    are backwards compatible, according to Avro's schema resolution rules. For
+    example, Materialize could choose to add a new field to the top-level record
+    containing additional metadata. This is a backwards compatible change as
+    long as Materialize can publish a new schema to the registry containing the
+    new field. Downstream consumers using the old schema (as a reader schema)
+    will simply discard the new field.
+
+    We'd not be able to evolve schemas like this if users provide Materialized
+    with the schema to use. The user-provided schema would not contain the new
+    fields, and Materialize would not have the authority to edit that schema to
+    contain the new fields.
+
+At a high level, allowing users to specify their desired Avro schema would make
+Materialize responsible for reconciling that Avro schema with the shape of the
+materialized view provided to the sink. This reconciliation would be quite
+delicate and difficult, and liable to cause sink breakage during version
+upgrades.
+
+So, as a philosophy, we propose that Materialize is always solely responsible
+for generating the Avro schema used by an Avro-formatted sink. Customizations to
+the generated Avro schema must be made via purpose-built options in the `FORMAT
+AVRO` clause.
+
+### Avro schema node specifiers
```

Comment:
Why is this section still needed if we're using SQL names instead?

Reply:
It makes sense to keep it under "Alternatives" still?